### ISSUE TITLE
support Farcaster

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#447]
A "sufficiently decentralized social network." Don't know much about it yet otherwise. Looks pretty heavy on crypto people, like Nostr (#446).

- https://www.farcaster.xyz/docs/intro
- https://github.com/a16z/awesome-farcaster
- https://paragraph.xyz/@kc/atproto-vs-farcaster

**Comment by snarfed** [comment]
Tried to create an account briefly, but eventually gave up. The [official client is invite only](https://warpcast.com/~/download), [this other client requires a crypto wallet](https://connect.phrasetown.com/), and [this other other client evidently can't create accounts](https://omilos.xyz/).

**Comment by snarfed** [comment]
Data point to help determine priority: https://farcaster.network/ says right now there are 11k users, 3.2k MAUs.

**Comment by snarfed** [comment]
Evidently https://petar.substack.com/ / https://staging.bsky.app/profile/petar.bsky.social is offering invitations to WarpCast.

**Comment by snarfed** [comment]
https://warpcast.com/acidbrn (@acidbrn?) generously sent me an invitation. Thank you Aaron!

**Comment by snarfed** [comment]
This has grown a lot in the last year! Looks like ~20x from https://farcaster.network/ , now 300k total users, 50-60k MAU.

**Comment by snarfed** [comment]
Huh. https://farcaster.network/ seems to have regressed, it's now showing data from back in Sept. 😕

**Comment by snarfed** [comment]
https://dune.com/pixelhack/farcaster seems current. 260k total users, 25k DAU. No MAU number.

**Comment by snarfed** [comment]
I've been looking at the Farcaster protocol a bit more. It's similar to Nostr, in that users own their identities and keys, and can use different clients, apps, and relays ("hubs") interchangeably.

Key differences:
* [Identity is blockchain-based, on Ethereum.](https://docs.farcaster.xyz/learn/what-is-farcaster/accounts) Creating a new account costs money.
  
  > You'll need to get a username, rent storage and add a key before you can use your account. These steps require many signatures and onchain transactions and can be tedious with a regular Ethereum wallet. We recommend starting with [Warpcast](https://www.warpcast.com/), a special Farcaster wallet which will handle the entire flow for you.
* [Human-friendly names are based on ENS.](https://docs.farcaster.xyz/learn/architecture/ens-names) Off-chain names, aka "fnames," are provided free by the Farcaster company. Alternatively, you can bring your own [ENS domain](https://ens.domains/), which costs money, but is independent of Farcaster the company.
* [Hubs](https://docs.farcaster.xyz/learn/architecture/hubs) run a gossip protocol to proactively synchronize data between each other and provide eventual consistency. Unlike Nostr, this means that it doesn't matter nearly as much which hub(s) you connect to as a user or client, nor do you need per-user relay hints, etc.

Also see https://neynar.com/ , popular third party Farcaster SDK.

**Comment by snarfed** [comment]
[Creating an account looks a bit intimidating and blockchain heavy](https://docs.farcaster.xyz/developers/guides/accounts/create-account), and [afaict costs $5](https://techcrunch.com/2024/02/06/farcaster-decentralized-social-network-mass-adoption/#us-tc-ros-mw-mid-center3), either one time [or per year](https://cryptopotato.com/farcaster-guide/), with [another $9/yr minimum for storage](https://cryptopotato.com/farcaster-guide/). Ugh. I wouldn't plan to subsidize that myself for people who want to bridge into Farcaster from other networks, and if they have to make a normal Farcaster account themselves first and then give Bridgy Fed the keys, that's a lot of friction. Hrm.

**Comment by snarfed** [comment]
Oof, looks like Farcaster is shrinking, [DAUs and interaction volume are evidently down by half (!) over the last 6 mos](https://dune.com/pixelhack/farcaster).

![image](https://github.com/user-attachments/assets/424629b8-3cf3-4d5a-87a7-cecf3dd778d9)


DAU/MAU ratio was roughly the same 6 mos ago as it is now, ~.275, so MAUs are probably down by half too.

![image](https://github.com/user-attachments/assets/8ecbbe76-76c7-4dce-9562-0b00c18610de)

**Comment by snarfed** [comment]
Talked with someone who works with Farcaster at the SF ATProto meetup yesterday. He pointed me to [snapchain](https://snapchain-docs.vercel.app/), their new version of hub synching that evidently replaces [deltagraph](https://www.varunsrinivasan.com/2024/04/28/the-goldilocks-consensus-problem), the old CRDT gossip based system.

**Comment by snarfed** [comment]
First 7m of this talk is a good summary of the Farcaster team's priorities as of May: 
https://www.youtube.com/watch?v=Q9eU1yxk-TI

TLDR:
* top priority is growth, of DAUs *with wallets*
* Doubling down on crypto/payments as differentiator
* Making it free for new users! ("apps, bring us your users" 👀👀👀)
* "Rewards," ie they pay people for usage, and they're ramping up
* Still lots of energy around frames, now called Mini Apps (basically frames 2.0)

**Comment by snarfed** [comment]
Start here: https://snapchain.farcaster.xyz/guides/writing-messages

Debugging tools: 
* https://explorer.neynar.com/
* https://casterscan.com/

Python lib (unmaintained): https://github.com/a16z/farcaster-py ... but uses the Warpcast API, which is deprecated, https://github.com/a16z/farcaster-py/issues/519 . We want to talk to hubs directly.

More: https://github.com/a16z/awesome-farcaster

**Comment by snarfed** [comment]
Is there a directory of Farcaster (snapchain) hubs? Haven't found one yet.

**Comment by snarfed** [comment]
Seems like clients and services, _if_ they talk directly to hubs via the hub APIs - which is maybe less common right now but is the direction the team and ecosystem want to head toward - generally hard-code a specific hub hostname. A few big ones are:
* ~~the team's official hub, `nemes.farcaster.xyz:2283`~~ ~~`hoyt.farcaster.xyz`~~ ~~`lamia.farcaster.xyz`~~ `crackle.farcaster.xyz`
* Neynar's, `hub-grpc-api.neynar.com`/`snapchain-grpc-api.neynar.com`, paid (?)
* Piñata's, `hub-grpc.pinata.cloud`, seems down?
  ```
  $ grpcurl --proto rpc.proto -format json hub-grpc.pinata.cloud:3383 HubService/GetInfo
  Failed to dial target host "hub-grpc.pinata.cloud:3383": tls: failed to verify certificate: x509: “CloudFlare Origin Certificate” certificate is not standards compliant
  ```
* ~~[Standard Crypto's](https://www.standardcrypto.vc/), `hub.farcaster.standardcrypto.vc`~~, seems down

Hubs have a mechanism for gossiping known hubs' hostname to each other, via ContactInfo messages, but maybe these aren't meant for clients? Not sure.

Snapchain hubs officially support both an HTTP API and a gRPC API. The gRPC one is more full featured and has the event subscription API (firehose) and inter-hub sync API.

(The deprecated Warpcast HTTP API was a different thing.)

**Comment by snarfed** [comment]
~TODO: read https://github.com/farcasterxyz/protocol/discussions/~

**Comment by snarfed** [comment]
https://dune.com/neynarxyz/farcaster

**Comment by snarfed** [comment]
Damn. I'd planned to use our usual domain-style handle as the Farcaster username (fname), but evidently fnames can only be 16 chars, no periods. Hrm. https://github.com/farcasterxyz/protocol/blob/main/docs%2FSPECIFICATION.md#user-content-name-server

**Comment by snarfed** [comment]
From Jan: https://neynar.com/blog/neynar-is-acquiring-farcaster

**Comment by snarfed** [comment]
[Snapchain Signers (FIP-266)](https://github.com/farcasterxyz/protocol/discussions/266) looks promising, will let us register new keys (eg for new users) off-chain!

Generating a new FID still has to be on chain afaict, so maybe not a big win, but maybe still something.

**Comment by snarfed** [comment]
It's alive!
⁂ [@snarfed@indieweb.social](https://indieweb.social/@snarfed) => 🏛️ [snarfed2](https://farcaster.xyz/snarfed2)
🏛️ [snarfed3](https://farcaster.xyz/snarfed3) => 🦋 [snarfed.eth.fc.brid.gy](https://bsky.app/profile/snarfed3.fc.brid.gy), ⁂ @[snarfed3@fc.brid.gy](mailto:snarfed3@fc.brid.gy), 🦩 [snarfed3@fc.brid.gy](https://coracle.social/nprofile1qqs0kqmqmjucsj5urzz58m36ghc7ddh49qetykrfaxs03ufxeqr06lcpp4mhxue69uhkummn9ekx7mqae43fx)

**Comment by snarfed** [comment]
Ruh roh. RIP Farcaster? Naynar is throwing in the towel. https://farcaster.xyz/rish/0xad8515c3 , https://farcaster.xyz/rish/0x049381ef
